### DISCUSSION TITLE
Newby question, help needed to make copy of DB to local

### DISCUSSION THREAD

**Discussion by AntonyNaude** [dbeaver/dbeaver#12312] [Q&A] [answered]
I am fairly new to dbeaver and I can basically run SQL statements to manage data, etc. 
I am helping a local owl sanctuary volunteer org with their database and I need to cleanup some incorrectly captured data.
I want to copy a complete PostgreSQL database with tables and all, to local, so that I can test my SQL queries, etc. without causing an issues to the existing online database. 
Please help.

**Comment by sknextcode** [comment]
In dbeaver, just right click on the database and select Tools > Dump database. This will open the wizard to dump your database to a local .backup file. You can then use this file on your local database with right click on database again > Tools > restore database.
If you have selected all items in the dump wizard, dbeaver dumps all of the table structure, table data, views, functions, sequences, etc.. Only thing that will be missing are postgreSQL extensions. But you can look them up under extensions in the productive database and install them on your local database using CREATE EXTENSION  (https://www.postgresql.org/docs/9.1/sql-createextension.html) if needed.

Be aware that we currently experiencing some problems when restoring with latest dbeaver version ( 21.0.3.202104181339 ) and windows 10. So you might have to use the command line, an older Version of dbeaver or another tool to restore the database on your local setup for now.

**Reply by AntonyNaude** [reply]
Hi, thank you for the response. I don't have "dump database" if I right click on database, tools, though.
I only have Open Dashboard, Session Manager, Lock Manager, create new task...

**Reply by sknextcode** [reply]
No problem. I guess, you have tried right clicking on the connection, rather than the database?
To verify: Double click on the connection and you should be able to see the databases. Then right click on the database that you want to dump/backup:
![image](https://user-images.githubusercontent.com/81296681/116693863-bf1a8a00-a9be-11eb-8213-fa304fb744f7.png)
In the above image "data" and "postgres" are the databases while "PostgreSQL - LOCAL" ist the connection.

**Reply by AntonyNaude** [reply]
I think this is correct?
![image](https://user-images.githubusercontent.com/83452468/116695887-7c0de600-a9c1-11eb-91e8-d8201b194a29.png),

**Reply by AntonyNaude** [reply]
datacleanup is the database

**Reply by AntonyNaude** [reply]
ok, if I click on Create new task, I get a list and it seems I can create PostgreSQl dump
![image](https://user-images.githubusercontent.com/83452468/116696197-e161d700-a9c1-11eb-8b7a-a148753766d5.png)

**Reply by AntonyNaude** [reply]
and this did dump, but I am not sure now if this is the correct method though

**Reply by AntonyNaude** [reply]
But how now do I restore it to local? Do I create a local first? I ask this because I keep getting errors creating a local or even a database.

**Reply by AntonyNaude** [reply]
![image](https://user-images.githubusercontent.com/83452468/116696796-a7dd9b80-a9c2-11eb-9df4-440258ec62d8.png)

**Reply by AntonyNaude** [reply]
![image](https://user-images.githubusercontent.com/83452468/116696991-e6735600-a9c2-11eb-9c95-80b73b98b2b0.png)

**Reply by sknextcode** [reply]
Ok, so it is labeled "Backup" not "Dump Database" in your context menu. 

![image](https://user-images.githubusercontent.com/81296681/116698516-c17fe280-a9c4-11eb-9ac9-d948d4376af1.png)


To restore it to local, first set up a local Database (you will need a local PostgreSQL database up and running).

**Reply by AntonyNaude** [reply]
Backup, got it. But now I'm failing at creating a local DB, pls see screen dumps above

**Reply by sknextcode** [reply]
Okay, so first of all you need to setup a local DB server (e.g. take a look at https://www.postgresqltutorial.com/install-postgresql/ on how to setup postgresql locally).
After that add a new connection to your local server in dbeaver, create a database and restore the backup into that database.

**Reply by sknextcode** [reply]
The official Download page is: https://www.postgresql.org/download/ but will redirect you to EDB anyway. So both ways will work the same.

**Reply by AntonyNaude** [reply]
thank you very much for your assistance, I really appreciate it.

**Reply by AntonyNaude** [reply]
mmm. looks like it only imported the database structure, no data.

**Comment by sknextcode** [comment]
Hi Antony,I'm currently not at my computer, but want to inform you, that as mentioned in my first reply, restoring with current version of dbeaver is not working. Please try an older version (7.3.5).We have reported this in the issues section aswell.As soon as I return home, I will try explaining another way to restore your dump file.

**Reply by AntonyNaude** [reply]
thank you

**Reply by sknextcode** [reply]
You're welcome. 
So we have two other options to complete our task as long as we cannot use dbeaver to restore our database.

1.  using pgAdmin to restore the database

While installing postgresql on your machine in order to create a local database instance, you may have selected the option to install pgAdmin aswell:
![image](https://user-images.githubusercontent.com/81296681/116730205-81325b80-a9e8-11eb-95b5-d3db40952a2d.png)
In above image it is the second option from top.

If so, just start up pgAdmin on your machine and connect to your local server (localhost) using the password you've provided during installation.
-> You should be able to right click the local database and choose "Restore.."
->Format:  leave the format at default ("Custom or tar")
-> Filename: Choose the .backup file you've created earlier
![image](https://user-images.githubusercontent.com/81296681/116730830-3ebd4e80-a9e9-11eb-924f-f30f6feed4e5.png)
-> Hit "Restore"
-> You can see the progress at the bottom right
![image](https://user-images.githubusercontent.com/81296681/116731016-7cba7280-a9e9-11eb-9638-09ff65b6f75f.png)
-> If it ends up with status "Failed (exit code: 1)." don't worry, that often relates to the schema public already existing. As long as it says "errors ignored on restore: 1" you are still good to go.
![image](https://user-images.githubusercontent.com/81296681/116731356-df137300-a9e9-11eb-979f-89debaf80263.png)
If you scroll to the top of the log you should see something similiar to:
![image](https://user-images.githubusercontent.com/81296681/116731240-bd19f080-a9e9-11eb-850a-6797c938dd18.png)
-> Just close the log and take a look at the data. It should be there.

2. using command line 

-> Assuming you are using a windows pc, just open up cmd.exe
-> navigate to the folder where your backup file is in (using cd command (e.g. if directly in drive C: `cd c:\`))
-> use the statement `pg_restore -U YOUR_USERNAME --password -h YOUR_SERVER -p YOUR_PORT -d YOUR_DATABASE_NAME YOUR_BACKUP_FILE`
(eg.: `pg_restore -U postgres --password -h localhost -p 5433 -d test_github test_full.backup`)
-> Hit enter (--pasword parameter means it should ask for password next)
-> Enter the password you have set for your local database
-> It will take a while
-> Remember what I've said about the possible error if schema public already exists? It can occur at this point aswell:
![image](https://user-images.githubusercontent.com/81296681/116732932-b68c7880-a9eb-11eb-822d-8fc5ced680be.png)
But as mentioned before, this is nothing to worry about. pg_restore will handle this for you.
-> And now: You are good to go.

**Reply by AntonyNaude** [reply]
I am going to try this in the morning, it is rather late here, will report back on progress

**Reply by AntonyNaude** [reply]
Hi Sebastian, I used pgAdmin and it seems it "worked" in restoring the data however on the source DB some of the tables have foreign keys, etc. which did not pull through. I am going to try again.
ps the backup file has a sql extension.

**Reply by sknextcode** [reply]
Hi Antony, first of all, file extension .sql should be fine, as I just like to set it to .backup to tell apart, what ist a script and what is a backup. Do you get any other errors than the one I mentioned in your pgAdmin log after restoring?

**Reply by AntonyNaude** [reply]
yes, there was a few, but somehow it seems "ok"

**Reply by AntonyNaude** [reply]
the error reads "pg_restore: error: could not execute query: ERROR:  role "data-user" does not exist"

**Comment by sknextcode** [comment]
Okay. So the role "data-user" exists on the remote database, but not your local one. 
You can try to add the parameters --no-privileges and --no-owner to pg_restore in order to avoid this.
In pgAdmin this should be available in the options tab of the restore wizard. Just slide Owner and Privilige in "Do not save" to "Yes".

**Reply by sknextcode** [reply]
![image](https://user-images.githubusercontent.com/81296681/116789505-85757c00-aaaf-11eb-9cac-48581d2d5234.png)

**Reply by AntonyNaude** [reply]
Thank, I am going to try this.
Just another question on creating the test DB, should I do it in Dbeaver or pgAdmin?, then import via pgAdmin.
Also should I wish to purge the test db, and rerun the import, or should I delete the db?

**Reply by AntonyNaude** [reply]
ps, setting Owner and Privilige in "Do not save" to "Yes" worked perfectly, now I only have the expected error, "pg_restore: error: could not execute query: ERROR:  schema "public" already exists"

**Reply by sknextcode** [reply]
You can use both tools. It should not make a difference.
Since the backup contains the whole database, I suggest deleting the test database an creating a new one to rerun the restore procedure.
If you are still having trouble with the foreign keys: Take a look at the log if it shows any more errors, especially regarding foreign keys.
Another question is: Are those foreign key structures necessary for your work on cleaning up the data? (e.g. is data you need to work on relying on them?) If so, and if the problems with the foreign keys persist, you could try adding them manually after restore. This may save time over troubleshooting these issues.

**Reply by sknextcode** [reply]
> ps, setting Owner and Privilige in "Do not save" to "Yes" worked perfectly, now I only have the expected error, "pg_restore: error: could not execute query: ERROR: schema "public" already exists"

Great news!

**Reply by AntonyNaude** [reply]
the foreign keys issues is solved

**Reply by sknextcode** [reply, marked as answer]
Very glad to hear that. So you can finally start working on the data, I guess. 

Briefly summarized:

1. you've used DBeaver to make a backup of the remote database
2. installed Postgres to run a local database server
3. created a local database using DBeaver/pgAdmin
4. Restored the backup into your local database using pgAdmin (since DBeaver currently has an issue wiht that, see https://github.com/dbeaver/dbeaver/issues/12293 for more information) and disabled owner and privileges upon restore (you can do so during backup aswell, but personally I would recommend choosing this on restore if needed)

**Reply by AntonyNaude** [reply]
Yes, big thanks for your help, I actually learned a lot.
I have deleted the test db a few times, recreated it, and imported the remote data. it works perfectly.

**Reply by sknextcode** [reply]
You're welcome. Glad, that I could help you getting started.

**Reply by venpisey12** [reply]
0100-05915377-12

**Comment by venpisey12** [comment]
0100-05915377-12
